### PR TITLE
feat: add existing vpc functionality

### PR DESCRIPTION
**Pull request by Aayush-Abhyarthi**
### Description

https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/issues/599

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: examples/default/variables.tf [hunk 1]
```diff
@@ -16,6 +16,11 @@ variable "prefix" {
   default     = "test-landing-zone"
 }
 
+variable "is_vpc_existing" {
+  description = "Indicates whether user wants to use an existing vpc or create a new one."
+  type        = bool
+}
+
```
Comment (shemau) [inline examples/default/variables.tf:23]: Based partially on the COS module as example.

This support almost certainly requires two new properties (to avoid the terraform feature about using values created at apply time in counts). It likely also requires an addition example to validate using existing values.

For variables to be consistent with the COS module (there may be other conflicting examples you would prefer to follow)

existing_cos_instance_id and create_cos_instance become existing_vpc_id (string) and create_vpc (bool), extending to existing_subnet_ids (map(string)) and create_subnets (bool)... etc.

File: main.tf [hunk 2]
```diff
@@ -24,6 +25,11 @@ resource "ibm_is_vpc" "vpc" {
   no_sg_acl_rules             = var.clean_default_sg_acl
 }
 
+data "ibm_is_vpc" "vpc" {
```
Comment (shemau) [inline main.tf:28]: A local will be required to determine which ibm_is_vpc.vpc is required

```hcl
locals {
  vpc_id = var.create_vpc ? var.existing_vpc_id : ibm_is_vpc.vpc.id
}
```

IF it is only the ID that is used, there may be no need to data resource lookup.

Additional validation is required to detect errors early (plan), so make sure that if create_vpc is false, then existing_vp_id MUST be passed.

**Review by shemau — DISMISSED** [review]
The landing-zone-vsi module already ONLY works on an existing VPC and existing subnets.

These feels like it might not be the right place for this code, it seems that calling a module that exists to create a VPC with an existing VPC is wrong.

Discuss at a deep dive, but maybe the features are required in one or more of the profiles or root module in landing-zone to avoid this module being called in the first place?

**Comment by ocofaigh** [comment]
@Aayush-Abhyarthi We had a call today about this. What we are going to try implement is the ability to pass an **existing VPC** to the module, and alternatively the ability to pass an **existing VPC and existing subnets**.
We should add an example to test this use case, but one thing you need to update in the code is the outputs. Ensure that all of the outputs work when using existing VPC and subnets. Currently the outputs will only work when creating new VPC.

File: variables.tf [hunk 3]
```diff
@@ -3,10 +3,22 @@
 ##############################################################################
 
 variable "name" {
-  description = "Name for VPC"
+  description = "Name for VPC."
   type        = string
 }
 
+variable "create_vpc" {
```
Comment (daniel-butler-irl) [inline variables.tf:10]: I don't think we need `create_vpc` and `existing_vpc_id `. Just do if `existing_vpc_id ` is null create a new vpc. @ocofaigh do you agree?

Comment (ocofaigh) [reply variables.tf:10]: We do need it - we will face issues where if you install the VPC in same terraform apply, this will fail saying it can't compute the plan on an unknown value. so hence it needs a boolean (we found this out the hard way before)

File: examples/existing_vpc/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Basic Example
```
Comment (ocofaigh) [inline examples/existing_vpc/README.md:1]: There is already a "Basic Example". I would rename this to "Existing VPC and subnets example"

File: examples/existing_vpc/main.tf [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+module "slz_vpc" {
+  source            = "../../"
+  resource_group_id = module.resource_group.resource_group_id
+  region            = var.region
+  tags              = var.resource_tags
+  create_vpc        = false
+  existing_vpc_id   = resource.ibm_is_vpc.example_vpc.id
+  create_subnets    = false
+
+  subnets = {
```
Comment (ocofaigh) [inline examples/existing_vpc/main.tf:32]: You have set `create_subnets` to `false` here, so I expect to see `existing_subnet_ids` to be used, so you will also need to create subnets in the "example_vpc" VPC before calling this module.

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@Aayush-Abhyarthi I think your updates to the outputs are wrong. You changed `vpc_name` and `vpc_crn` outputs to now use the `data.ibm_is_vpc` data lookup, however, this data lookup has a count on it that means it only executes if `create_vpc = false`. In order to ensure there are no regressions introduced here with the outputs, I think you need to add ALL outputs to all examples that the tests run, and add logic to all the tests to ensure each of the outputs has a value.

File: variables.tf [hunk 6]
```diff
@@ -3,8 +3,21 @@
 ##############################################################################
 
 variable "name" {
-  description = "Name for VPC"
+  description = "Name for VPC."
```
Comment (ocofaigh) [inline variables.tf:6]: Perhaps update to: "The name to give the newly provisioned VPC. Only used if 'create_vpc' is true."

File: variables.tf [hunk 7]
```diff
@@ -3,8 +3,21 @@
...
   type        = string
+  default     = null
```
Comment (ocofaigh) [inline variables.tf:8]: hmm, by setting the default value to null, it means the current set of the module variable default values won't work out of the box (since `create_vpc` is `true` by default, meaning `name` is required). Perhaps we default the value to some real string like "dev" ?

File: variables.tf [hunk 8]
```diff
@@ -20,6 +33,7 @@ variable "region" {
 variable "prefix" {
   description = "The prefix that you would like to append to your resources. Explicitly set to null if you do not wish to use a prefix."
```
Comment (ocofaigh) [inline variables.tf:34]: Just noticed the word "append" here would indicate the value is added to the end of the resource name, which is not true. Suggest rewording to: "The value that you would like to prefix to the name of the resources provisioned by this module. Explicitly set to null if you do not wish to use a prefix."

File: .ipynb_checkpoints/Untitled-checkpoint.ipynb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
+ "cells": [],
+ "metadata": {},
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Comment (ocofaigh) [inline .ipynb_checkpoints/Untitled-checkpoint.ipynb:6]: what is the file, was it accidentally committed?

File: Untitled.ipynb [hunk 10]
```diff
@@ -0,0 +1,98 @@
+{
```
Comment (ocofaigh) [inline Untitled.ipynb:1]: what is the file, was it accidentally committed?

File: default_security_group.tf [hunk 11]
```diff
@@ -12,7 +12,7 @@ locals {
 
 resource "ibm_is_security_group_rule" "default_vpc_rule" {
   for_each  = local.security_group_rule_object
-  group     = ibm_is_vpc.vpc.default_security_group
+  group     = ibm_is_vpc.vpc[0].default_security_group
```
Comment (ocofaigh) [inline default_security_group.tf:15]: Hmm, this means this code block will only occur if a new VPC is created. Won't we want it to happen on an existing VPC too if one is passed ? @toddgiguere @digvijay-ukirde ?

File: examples/existing_vpc/README.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+# Existing VPC and subnets Example
+
+A very basic example to provision SLZ VPC with existing vpc.
```
Comment (ocofaigh) [inline examples/existing_vpc/README.md:3]: It's not really a basic example. It's an example of calling the module with an existing VPC and subnets

File: examples/existing_vpc/main.tf [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+module "resource_group" {
+  source  = "terraform-ibm-modules/resource-group/ibm"
+  version = "1.1.0"
+  # if an existing resource group is not set (null) create a new one using prefix
+  resource_group_name          = var.resource_group == null ? "${var.prefix}-resource-group" : null
+  existing_resource_group_name = var.resource_group
+}
+
+#############################################################################
+# Provision VPC
```
Comment (ocofaigh) [inline examples/existing_vpc/main.tf:14]: It probably doesn't make sense to provision the VPC and subnets in the example here, as this is not something someone using existing vpc + subnets resources would be doing. We could maybe have terraform code in a different folder that provisions the VPC + subnets, and have our test call that code first, and pass the VPC + subnet details to this example. We did this in the [TestRunExistingGroupExample](https://github.com/terraform-ibm-modules/terraform-ibm-resource-group/blob/0058f23ed3bab77970a20d5791d4bb705fc28a95/tests/pr_test.go#L41C6-L41C33) test if you want to use it for reference. 
Suggest getting the rest of the code up to date an working first before updating this though.

File: examples/existing_vpc/main.tf [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+module "slz_vpc" {
+  source              = "../../"
+  resource_group_id   = module.resource_group.resource_group_id
+  region              = var.region
+  tags                = var.resource_tags
+  create_vpc          = false
+  existing_vpc_id     = resource.ibm_is_vpc.example_vpc.id
+  create_subnets      = false
+  existing_subnet_ids = rersource.ibm_is_subnet.testacc_subnet.id
```
Comment (ocofaigh) [inline examples/existing_vpc/main.tf:39]: how does this make sense? The `existing_subnet_ids` is a map value, yet you are passing a single string to it?

File: examples/existing_vpc/main.tf [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+
+  subnets = {
```
Comment (ocofaigh) [inline examples/existing_vpc/main.tf:42]: I don't see this being used anywhere?

File: examples/existing_vpc/output.tf [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+output "vpc_id" {
+  value       = module.slz_vpc.vpc_id
+  description = "VPC id"
+}
+
+output "vpc_crn" {
+  value       = module.slz_vpc.vpc_crn
+  description = "VPC crn"
+}
```
Comment (ocofaigh) [inline examples/existing_vpc/output.tf:13]: In order to ensure we have the outputs correct for this use case, can you output everything that the root level module outputs so we can check all values in the test. Thanks

File: examples/existing_vpc/variables.tf [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+variable "region" {
+  description = "The region to which to deploy the VPC"
+  type        = string
+  default     = "us-south"
+}
+
+variable "prefix" {
+  description = "The prefix that you would like to append to your resources"
+  type        = string
+  default     = "basic-slz-vpc"
```
Comment (ocofaigh) [inline examples/existing_vpc/variables.tf:16]: `basic-slz-vpc` -> `existing-slz-vpc`

File: examples/existing_vpc/variables.tf [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+variable "name" {
+  description = "The name of the vpc"
+  type        = string
+  default     = "vpc"
+}
```
Comment (ocofaigh) [inline examples/existing_vpc/variables.tf:23]: Name won't be needed if using an existing VPC

File: main.tf [hunk 19]
```diff
@@ -59,6 +66,15 @@ resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_crn" {
   }
 }
 
+data "ibm_is_vpc" "vpc" {
+  count = var.create_vpc == false ? 1 : 0
+  id    = var.existing_vpc_id
```
Comment (ocofaigh) [inline main.tf:71]: I think you will need to specify the resource group here, meaning you need a new inout variable called `existing_resource_group_id`

File: outputs.tf [hunk 20]
```diff
@@ -4,17 +4,17 @@
 
 output "vpc_name" {
   description = "Name of VPC created"
-  value       = ibm_is_vpc.vpc.name
+  value       = var.create_vpc == true ? var.name : data.ibm_is_vpc.vpc[0].name
```
Comment (ocofaigh) [inline outputs.tf:7]: You should not use variables in outputs like this. Always pull a value from the resource itself. So change `var.name` -> `ibm_is_vpc[0].vpc.name`

File: outputs.tf [hunk 21]
```diff
@@ -4,17 +4,17 @@
...
 output "vpc_id" {
   description = "ID of VPC created"
-  value       = ibm_is_vpc.vpc.id
+  value       = local.vpc_id
 }
 
 output "vpc_crn" {
   description = "CRN of VPC created"
-  value       = ibm_is_vpc.vpc.crn
+  value       = var.create_vpc == true ? resource.ibm_is_vpc.vpc[0].crn : data.ibm_is_vpc.vpc[0].crn
```
Comment (ocofaigh) [inline outputs.tf:17]: Does this `resource.ibm_is_vpc.vpc[0].crn` work?? I would usually expect just `ibm_is_vpc.vpc[0].crn`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@Aayush-Abhyarthi Still alot of issues with this PR. I left some comments. Please reach out if you are unclear on anything

File: main.tf (hunk 19 above)
Comment (Aashiq-J) [reply main.tf:121]: There is no mention of `resource_group` being a parameter for data source
https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/is_vpc

File: variables.tf [hunk 22]
```diff
@@ -18,8 +31,9 @@ variable "region" {
 }
 
 variable "prefix" {
-  description = "The prefix that you would like to append to your resources. Explicitly set to null if you do not wish to use a prefix."
+  description = "The value that you would like to prefix to the name of the resources provisioned by this module. Explicitly set to null if you do not wish to use a prefix."
   type        = string
+  default     = null
```
Comment (ocofaigh) [inline variables.tf:36]: thoughts on maybe adding a default prefix value here instead of `null` to be consistent with `name` variable? Or should we default `name` to null? cc @toddgiguere

File: tests/existing-resources/README.md [hunk 23]
```diff
@@ -0,0 +1,3 @@
+# Existing Resources For use in tests
+
+<!-- Add an example where existing resources can be passed as variables for local testing (not for use with tests) -->
```
Comment (ocofaigh) [inline tests/existing-resources/README.md:3]: `not for use with tests` is confusing - we are using it for tests, but as a prereq deployment to test the existing resources example

**Review by ocofaigh — COMMENTED** [review]
This is looking much better. I left some final minor comments. @toddgiguere would also like to get a second pari of eyes on it please

File: tests/existing-resources/README.md [hunk 24]
```diff
@@ -0,0 +1,3 @@
+# Existing Resources For use in tests
```
Comment (SirSpidey) [inline tests/existing-resources/README.md:1]:
```suggestion
# Existing resources to use in tests
```

Comment (SirSpidey) [reply tests/existing-resources/README.md]: Just a capitalization comment. See Conall's comments about what the title should be.

File: variables.tf (hunk 22 above)
Comment (Aashiq-J) [reply variables.tf:36]: we will need some more code modification to support adding default value for `prefix`, while adding a default value for `name` works fine.
Screenshot 2023-11-15 at 6 17 27 PM

Comment (ocofaigh) [reply variables.tf:36]: I was suggesting we either set `name` and `prefix` to `null`, or give both default string values

**Review by ocofaigh — COMMENTED** [review]
Just need to resolve the comments around what default value to give `name` and `prefix`

File: variables.tf (hunk 22 above)
Comment (Aashiq-J) [reply variables.tf:36]: If we are to set `name` and `prefix` to `null`, we will need to update the naming logic everywhere for the existing-vpc example to work.

**Comment by ocofaigh** [comment]
@toddgiguere would you mind giving this a final review please?

**Review by toddgiguere — APPROVED** [review]

**Review by ocofaigh — APPROVED** [review]
